in-progress/8509-prover-coordination/design.md

```diff
@@ -0,0 +1,124 @@
+
+The aztec node will provide an interface for proving marketplaces to submit quotes for proving an epoch.
+
+A proposer will be able to respond to a quote with a signed message, which a prover can submit to L1, along with a prover commitment bond. The prover can subsequently submit the proof to L1, releasing the bond.
+
+This coordination will take place "out of protocol", via an optional p2p topic, with the slight exception/leak that the rollup contract will permit the prover to submit bonds/proofs in a proposer's stead.
+
+## Introduction
+
+Presently, there is no coordination between proposers or provers. A coordination mechanism is necessary to ensure that:
+
+1. Proposers are able to obtain proofs for epochs, considering they will likely not have proving infrastructure.
+2. Provers are not obliged to "race" each other to submit proofs, as this would be a waste of resources.
+
+The protocol only "cares" about the proofs being submitted, but in practice, the node will need to coordinate the submission of proofs in a timely, efficient manner.
+
+## Interface
+
+Proving marketplaces will run full nodes, which will follow the pending chain. The node will expose a json-rpc endpoint for `submitEpochProofBid`, with the following signature:
+
+```typescript
+interface ProverCoordination {
+  submitEpochProofBid(epoch: number, basisPointFee: number): Promise<void>;
+}
+```
+
+This will be exposed via the cli as `aztec prover-coordination submit-epoch-proof-bid --epoch 123 --basis-point-fee 1000`.
+
+Under the hood, the node will submit the following message to `/aztec/proof-quotes/make/0.1.0` 
+
+
+```solidity
+struct Quote {
+  address prover;
+  uint256 epoch;
+  uint32 basisPointFee;
+  Signature signature;
+}
+```
+
+The `signature` will be produced using the L1 private key defined in the environment variable `PROVER_PUBLISHER_PRIVATE_KEY` to sign the message `keccak256(abi.encode(prover, epoch, basisPointFee))`.,
+
+Proposers will counter-sign the quote by submitting a message to `/aztec/proof-quotes/take/0.1.0` with the following same `Quote` struct, but with the `signature` produced using the L1 private key defined in the environment variable `VALIDATOR_PRIVATE_KEY`.
+
+The prover will be able to submit this Quote to `claimProofRight` on the rollup contract. The rollup contract will verify that:
+- The epoch is in the proof claim phase
+- There is not already a claim/proof for this epoch
+- The quote has been signed by the proposer for the current slot
+- The `prover` matches `msg.sender`
+- The epoch matches the previous epoch (i.e. the one to be proven)
+- The prover successfully transmits the bond (in TST) to the rollup contract
+
+If all conditions are met, the rollup contract stores the quote, and the address of the proposer.
+
+When the prover submits the proof, the rollup contract will pay out TST rewards to the proposer after paying the prover the basis point fee contained in the quote. It will also release the bond to the prover.
```

Comment:
Following on from what @spalladino says, it feels to me like the proposer should post the message to L1. That way, a signed message from a prover is a binding commitment that a proposer can use to post to L1. Currently it's not binding at all as it requires the prover to later post to L1 which they can simply choose not to creating a lot of uncertainty for the proposer.

Reply:
That is how I had originally been thinking of it.

I think the downside to this approach is that all provers need to `approve` the rollup contract to transfer funds ahead of time; the current approach does not. I don't think it is possible for the proposer to upload anything to the rollup and be able to transfer the prover's funds without prior action from the prover (assuming the TST contract does not trust the rollup). @LHerskind do you have any thoughts here?

Having the proposer submit though would also decrease a round of p2p, which would be huge.

Reply:
Why is it a problem to require a prior erc20 approval, or a pre-signed permit-like message to be included in the Quote sent by the prover? We could also force provers to escrow the funds for a period of time, to prevent them from cancelling the approval before the sequencer has had a chance to submit their quote.

Reply:
Not a serious problem, just would require more gas and coordination (once my bond is taken, i need to top up another approval). 

But I think the permit based flow is best!

I think an escrow step could be a nice extension in the future if cancelled approvals are a recurring problem.

Reply:
I previously shared thoughts via message but posting for visibility. 

Agree with previous comments that proposer should post the bond and submit the proof. If prover submits bond then:

1) They can DOS the system (no balance checks are being made prior to sending quote). 
2) Forces proposers to send quotes to multiple provers to hedge risk which may end up in less than optimal execution - basically the fastest prover to get to L1 wins the bid.

Reply:
I think the DOS vector is still present unless the proposer checks the L1 balances of the provers.

RE suboptimal, I think it would be "the fastest prover among those Alice approved wins the bid". I think that part is actually fine, the less desirable piece would be that provers who try to cash and fail to win the race condition in just wasted gas.

Reply:
Regarding DOS, you're right. But the escrow mechanism from above only works if the proposer posts the bond. 

Regarding suboptimal, because provers could ultimately decide to not post the bond, a proposer will be forced to approve more quotes than they would normally have if they had the ability to "exercise" a prover's bid. 

Because there's opportunity cost for the proposer in missing out on the proving rewards, their strategy might even be to approve all quotes from reputable provers driving the price down.

Reply:
Thank you @LHerskind! I updated the https://github.com/AztecProtocol/engineering-designs/pull/22 to talk about the escrow and permit flow.